'orphan parameter reference' when match type destructures a type alias with a bounded parameter

## Minimized code

```Scala
type Foo[A, B <: A] = A

type Bar[A] = A match {
  case Foo[String, x] => Unit
}
```

## Output (click arrow to expand)
<details>

```scala
error when pickling type A
error when pickling type  <: A
error when pickling tree x @ _
error when pickling tree Foo[String, x @ _]
error when pickling tree case Foo[String, x @ _] => Unit
error when pickling tree A match 
  {
    case Foo[String, x @ _] => Unit
  }
error when pickling tree [A >: Nothing <: Any] =>> 
  A match 
    {
      case Foo[String, x @ _] => Unit
    }
error when pickling tree type Bar[A >: Nothing <: Any] = 
  A match 
    {
      case Foo[String, x @ _] => Unit
    }
error when pickling tree () extends Object(), _root_.scala.Serializable { this: rs$line$2.type =>
  private def writeReplace(): AnyRef = 
    new scala.runtime.ModuleSerializationProxy(classOf[rs$line$2.type])
  type Bar[A >: Nothing <: Any] = 
    A match 
      {
        case Foo[String, x @ _] => Unit
      }
}
error when pickling tree @scala.annotation.internal.SourceFile("rs$line$2") final module class rs$line$2$
  (
) extends Object(), _root_.scala.Serializable { this: rs$line$2.type =>
  private def writeReplace(): AnyRef = 
    new scala.runtime.ModuleSerializationProxy(classOf[rs$line$2.type])
  type Bar[A >: Nothing <: Any] = 
    A match 
      {
        case Foo[String, x @ _] => Unit
      }
}
error when pickling tree package <empty> {
  final lazy module val rs$line$2: rs$line$2$ = new rs$line$2$()
  @scala.annotation.internal.SourceFile("rs$line$2") final module class 
    rs$line$2$
  () extends Object(), _root_.scala.Serializable { this: rs$line$2.type =>
    private def writeReplace(): AnyRef = 
      new scala.runtime.ModuleSerializationProxy(classOf[rs$line$2.type])
    type Bar[A >: Nothing <: Any] = 
      A match 
        {
          case Foo[String, x @ _] => Unit
        }
  }
}
Exception in thread "main" java.lang.AssertionError: assertion failed: orphan parameter reference: TypeParamRef(A)
	at dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:279)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:154)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$5(TreePickler.scala:246)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:248)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:154)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$20(TreePickler.scala:506)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:507)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$49$$anonfun$1(TreePickler.scala:588)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$31(TreePickler.scala:588)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:588)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$13(TreePickler.scala:478)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:478)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$50$$anonfun$1(TreePickler.scala:594)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$32(TreePickler.scala:594)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:595)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$34(TreePickler.scala:604)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:604)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTpt(TreePickler.scala:306)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$2(TreePickler.scala:321)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:325)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:536)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:344)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:344)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$26(TreePickler.scala:560)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:561)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$2(TreePickler.scala:320)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:325)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:536)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:344)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:344)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$28(TreePickler.scala:570)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:49)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:570)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:736)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:736)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$9$$anonfun$8(Pickler.scala:63)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:87)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:87)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:309)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:310)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:91)
	at dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:181)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
	at dotty.tools.dotc.Run.runPhases$5(Run.scala:191)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:199)
	at dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:64)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:206)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:148)
	at dotty.tools.repl.ReplCompiler.runCompilationUnit(ReplCompiler.scala:153)
	at dotty.tools.repl.ReplCompiler.compile(ReplCompiler.scala:163)
	at dotty.tools.repl.ReplDriver.compile(ReplDriver.scala:233)
	at dotty.tools.repl.ReplDriver.interpret(ReplDriver.scala:196)
	at dotty.tools.repl.ReplDriver.loop$1(ReplDriver.scala:129)
	at dotty.tools.repl.ReplDriver.runUntilQuit$$anonfun$1(ReplDriver.scala:132)
	at dotty.tools.repl.ReplDriver.withRedirectedOutput(ReplDriver.scala:151)
	at dotty.tools.repl.ReplDriver.runUntilQuit(ReplDriver.scala:132)
	at dotty.tools.repl.Main$.main(Main.scala:6)
	at dotty.tools.repl.Main.main(Main.scala)
```
</details>

A less minimized example

```scala
import scala.compiletime.ops.int.*

sealed trait HList
final case class HCons[H <: Int & Singleton, T <: HList](head: H, tail: T) extends HList
sealed trait HNil extends HList
case object HNil extends HNil

sealed trait Tensor[T, S <: HList]

object tf:
  def zeros[S <: HList](shape: S): Tensor[Float, S] = ???

  type NumElements[X <: HList] <: Int = X match
    case HNil => 1
    case HCons[head, tail] => head * NumElements[tail]

  def reshape[T, From <: HList, To <: HList](tensor: Tensor[T, From], shape: To)(using NumElements[From] =:= NumElements[To]): Tensor[T, To] = ???

object test:
  val x = HCons(1, HCons(2, HNil))
  val y = tf.reshape(tf.zeros(x), HCons(2, HCons(1, HNil)))
```

<details>
<summary>Error</summary>

```scala
error when pickling type head
error when pickling type HCons[head, tail]
error when pickling type Tensor[Float, HCons[head, tail]]
error when pickling tree Tensor[Float, HCons[head, tail]]
error when pickling tree val y: Tensor[Float, HCons[head, tail]] = 
  tf.reshape[Float, HCons[(1 : Int), HCons[(2 : Int), HNil.type]], 
    HCons[head, tail]
  ](tf.zeros[HCons[(1 : Int), HCons[(2 : Int), HNil.type]]](test.x), 
    HCons.apply[head, tail](2, HCons.apply[head, tail](1, HNil))
  )(<:<.refl[tf.NumElements[HCons[(1 : Int), HCons[(2 : Int), HNil.type]]]])
error when pickling tree () extends Object() { this: test.type =>
  private def writeReplace(): AnyRef = tal 0s
    new scala.runtime.ModuleSerializationProxy(classOf[test.type])
  val x: HCons[(1 : Int), HCons[(2 : Int), HNil.type]] = 
    HCons.apply[(1 : Int), HCons[(2 : Int), HNil.type]](1, 
      HCons.apply[(2 : Int), HNil.type](2, HNil)
    )
  val y: Tensor[Float, HCons[head, tail]] = 
    tf.reshape[Float, HCons[(1 : Int), HCons[(2 : Int), HNil.type]], 
      HCons[head, tail]
    ](tf.zeros[HCons[(1 : Int), HCons[(2 : Int), HNil.type]]](test.x), 
      HCons.apply[head, tail](2, HCons.apply[head, tail](1, HNil))
    )(<:<.refl[tf.NumElements[HCons[(1 : Int), HCons[(2 : Int), HNil.type]]]])
}
error when pickling tree @scala.annotation.internal.SourceFile(
  | => root / Compile / compileIncremental 0s
    "/home/maxime/code/dotty-bug-report/hlist-assertion-fail/src/main/scala/Main.scala"
  
) final module class test$() extends Object() { this: test.type =>
  private def writeReplace(): AnyRef = 
    new scala.runtime.ModuleSerializationProxy(classOf[test.type])
  val x: HCons[(1 : Int), HCons[(2 : Int), HNil.type]] = 
    HCons.apply[(1 : Int), HCons[(2 : Int), HNil.type]](1, 
      HCons.apply[(2 : Int), HNil.type](2, HNil)
    )
  val y: Tensor[Float, HCons[head, tail]] = 
    tf.reshape[Float, HCons[(1 : Int), HCons[(2 : Int), HNil.type]], 
      HCons[head, tail]
    ](tf.zeros[HCons[(1 : Int), HCons[(2 : Int), HNil.type]]](test.x), 
      HCons.apply[head, tail](2, HCons.apply[head, tail](1, HNil))
    )(<:<.refl[tf.NumElements[HCons[(1 : Int), HCons[(2 : Int), HNil.type]]]])
}
error when pickling tree package <empty> {
  import scala.compiletime.ops.package.int.*s
  final lazy module val test: test$ = new test$()
  @scala.annotation.internal.SourceFile(
    
      "/home/maxime/code/dotty-bug-report/hlist-assertion-fail/src/main/scala/Main.scala"
    
  ) final module class test$() extends Object() { this: test.type =>
    private def writeReplace(): AnyRef = 
      new scala.runtime.ModuleSerializationProxy(classOf[test.type])
    val x: HCons[(1 : Int), HCons[(2 : Int), HNil.type]] = 
      HCons.apply[(1 : Int), HCons[(2 : Int), HNil.type]](1, 
        HCons.apply[(2 : Int), HNil.type](2, HNil)
      )
    val y: Tensor[Float, HCons[head, tail]] = 
      tf.reshape[Float, HCons[(1 : Int), HCons[(2 : Int), HNil.type]], 
        HCons[head, tail]
      ](tf.zeros[HCons[(1 : Int), HCons[(2 : Int), HNil.type]]](test.x), 
        HCons.apply[head, tail](2, HCons.apply[head, tail](1, HNil))
      )(<:<.refl[tf.NumElements[HCons[(1 : Int), HCons[(2 : Int), HNil.type]]]])
  }
}
[info] exception occurred while compiling /home/maxime/code/dotty-bug-report/hlist-assertion-fail/src/main/scala/Main.scala
java.lang.AssertionError: assertion failed: orphan parameter reference: TypeParamRef(head) while compiling /home/maxime/code/dotty-bug-report/hlist-assertion-fail/src/main/scala/Main.scala
[error] ## Exception when compiling 1 sources to /home/maxime/code/dotty-bug-report/hlist-assertion-fail/target/scala-3.0.0-M1/classes
[error] java.lang.AssertionError: assertion failed: orphan parameter reference: TypeParamRef(head)
[error] dotty.DottyPredef$.assertFail(DottyPredef.scala:17)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:297)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:172)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$11$$anonfun$1(TreePickler.scala:189)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$1(TreePickler.scala:189)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:70)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:189)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:172)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$11$$anonfun$1(TreePickler.scala:189)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$1(TreePickler.scala:189)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:70)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:189)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:172)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:596)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTpt(TreePickler.scala:324)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$2(TreePickler.scala:340)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:70)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:344)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:547)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:368)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:368)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$26(TreePickler.scala:584)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:70)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:585)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$2(TreePickler.scala:339)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:70)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:344)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:560)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:368)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:368)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$28(TreePickler.scala:594)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:70)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:594)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:763)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:763)
[error] dotty.tools.dotc.transform.Pickler.run$$anonfun$3$$anonfun$2(Pickler.scala:69)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:105)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.transform.Pickler.run(Pickler.scala:105)
[error] dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:296)
[error] scala.collection.immutable.List.map(List.scala:246)
[error] dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:297)
[error] dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:110)
[error] dotty.tools.dotc.Run.runPhases$4$$anonfun$4(Run.scala:185)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] dotty.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$5(Run.scala:195)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:203)
[error] dotty.runtime.function.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:67)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:210)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:147)
[error] dotty.tools.dotc.Run.compile(Run.scala:129)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:38)
[error] dotty.tools.dotc.Driver.process(Driver.scala:194)
[error] dotty.tools.dotc.Main.process(Main.scala)
[error] xsbt.CachedCompilerImpl.run(CachedCompilerImpl.java:69)
[error] xsbt.CompilerInterface.run(CompilerInterface.java:41)
[error] sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] java.lang.reflect.Method.invoke(Method.java:498)
[error] sbt.internal.inc.AnalyzingCompiler.invoke(AnalyzingCompiler.scala:330)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:113)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:232)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:573)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:491)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2176)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2133)
[error] sbt.internal.io.Retry$.apply(Retry.scala:40)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2129)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] java.lang.Thread.run(Thread.java:748)
[error]            
[error] stack trace is suppressed; run last Compile / compileIncremental for the full output
[error] (Compile / compileIncremental) java.lang.AssertionError: assertion failed: orphan parameter reference: TypeParamRef(head)
```

</details>

--- a/tests/pos/i10242.scala
+++ b/tests/pos/i10242.scala
@@ -1,0 +1,32 @@
+// https://github.com/lampepfl/dotty/issues/10242  
+type Foo[A, B <: A] = A
+
+type Bar[A] = A match {
+  case Foo[String, x] => Unit
+}
+
+import scala.compiletime.ops.int.*
+
+sealed trait HList
+final case class HCons[H <: Int & Singleton, T <: HList](head: H, tail: T)
+    extends HList
+sealed trait HNil extends HList
+case object HNil extends HNil
+
+sealed trait Tensor[T, S <: HList]
+
+object tf:
+  def zeros[S <: HList](shape: S): Tensor[Float, S] = ???
+
+  type NumElements[X <: HList] <: Int = X match
+    case HNil              => 1
+    case HCons[head, tail] => head * NumElements[tail]
+
+  def reshape[T, From <: HList, To <: HList](
+      tensor: Tensor[T, From],
+      shape: To
+  )(using NumElements[From] =:= NumElements[To]): Tensor[T, To] = ???
+
+object test:
+  val x = HCons(1, HCons(2, HNil))
+  val y = tf.reshape(tf.zeros(x), HCons(2, HCons(1, HNil)))

